Source/JavaScriptCore/wasm/WasmTypeDefinition.h

```diff
@@ -94,8 +94,127 @@ enum Mutability : uint8_t {
     Immutable = 0
 };
 
-struct FieldType {
-    Type type;
+inline const char* makeString(PackedType packedType)
+{
+    switch (packedType) {
+    case PackedType::I8:
+        return "i8";
+    case PackedType::I16:
+        return "i16";
+    }
+    ASSERT_NOT_REACHED();
+}
+
+struct StorageType {
+public:
+    template <typename T>
+    bool is() const { return std::holds_alternative<T>(m_storageType); }
+
+    template <typename T>
+    const T* as() const { ASSERT(is<T>()); return std::get_if<T>(&m_storageType); }
+
+    StorageType() {
+        // FIXME: ???
+        m_storageType = std::variant<Type, PackedType>();
+    }
+
+    StorageType(Type t) {
+        m_storageType = std::variant<Type, PackedType>(t);
+    }
+
+    StorageType(PackedType t) {
+        m_storageType = std::variant<Type, PackedType>(t);
+    }
+
+    // If this is a type, returns itself; if it's a packed type, returns i32
+    Type unpacked() const {
+        if (is<Type>()) {
+            return *as<Type>();
+        }
+        return Types::I32;
+    }
+
+    size_t elementSize() const {
+        if (is<Type>()) {
+            switch (as<Type>()->kind) {
+            case Wasm::TypeKind::I32:
+            case Wasm::TypeKind::F32:
+                return sizeof(uint32_t);
+            default:
+                return sizeof(uint64_t);
+            }
+        }
+        switch (*as<PackedType>()) {
+        case PackedType::I8:
+            return sizeof(uint8_t);
+        case PackedType::I16:
+            return sizeof(uint16_t);
+        }
+        RELEASE_ASSERT_NOT_REACHED();
+    }
+
+    bool operator==(const StorageType& rhs) const {
+        if (rhs.is<PackedType>()) {
+            return (is<PackedType>() && as<PackedType>() == rhs.as<PackedType>());
+        }
+        if (!is<Type>()) {
+            return false;
+        }
+        return(as<Type>() == rhs.as<Type>());
+    }
+    bool operator!=(const StorageType& rhs) const { return !(*this == rhs); };
+
+    auto hash(uint32_t start) const {
+        unsigned accumulator = start;
+        uint8_t ty;
+        if (is<Type>()) {
+            ty = static_cast<uint8_t>(as<Type>()->kind);
+        } else {
+            switch (*as<PackedType>()) {
+            case PackedType::I8:
+                ty = JSC::Wasm::numTypes;
```

Comment:
I think it might be possible to simplify this. Since packed type codes are assigned to be distinct from regular type codes, it seems possible to just hash the actual packed type value (SLEB `-6` or `-7`) or the regular type code instead of taking a range at the end of the type codes.

e.g., you could define a `kind()` method on `StorageType` that returns either the packed type code or the regular type code, and then the hashing code in `WasmTypeDefinition.cpp` wouldn't have to change very much.